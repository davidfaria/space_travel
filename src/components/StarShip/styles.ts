import styled from 'styled-components'

export const Wrapper = styled.div`
  background: rgba(41, 41, 46, 0.9);
  border-color: rgb(41, 41, 46);
  max-width: 380px;
  width: 100%;
  height: 100px;
  padding: 16px;
  margin-bottom: 8px;
  border-radius: 4px;

  display: flex;
  justify-content: space-between;

  cursor: pointer;

  &:hover {
    background: rgb(38, 38, 43);
  }
`
export const ShipInfo = styled.div`
  display: flex;
  align-items: center;
`

export const Ship = styled.img`
  width: 32px;
  height: auto;
  margin-right: 8px;
`
export const Name = styled.strong`
  color: #fff;
  font-weight: bold;
  font-size: 1.6rem;
  text-transform: uppercase;
`

export const StopInfo = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

export const StopAmount = styled.span`
  color: #fff;
  font-size: 2rem;
`
export const StopBoard = styled.img`
  width: 48px;
  height: auto;
`
