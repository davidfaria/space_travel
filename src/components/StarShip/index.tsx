import React from 'react'
import * as S from './styles'
import { calculateStoped } from 'utils/calculateStop'

export type StarShipProps = {
  MGLT?: string
  cargo_capacity: string
  consumables: string
  cost_in_credits: string
  crew: string
  hyperdrive_rating: string
  length: string
  manufacturer: string
  max_atmosphering_speed: string
  name: string
  passengers: string
  starship_class: string
}

export type StartShipType = {
  starship: StarShipProps
  distance?: number
}

const StarShip = ({ starship, distance }: StartShipType) => {
  const stop_amount = React.useMemo(() => {
    if (!distance) return null
    return calculateStoped(
      distance,
      Number(starship.MGLT),
      starship.consumables
    )
  }, [starship, distance])

  return (
    <S.Wrapper>
      <S.ShipInfo>
        <S.Ship src="/img/rocket.png" loading="lazy" alt="Rocket Ship" />
        <S.Name>{starship.name}</S.Name>
      </S.ShipInfo>

      <S.StopInfo>
        <S.StopBoard src="/img/stop.png" loading="lazy" alt="Stop Soard" />
        <S.StopAmount>{stop_amount ? stop_amount : '*****'}</S.StopAmount>
      </S.StopInfo>
    </S.Wrapper>
  )
}

export default StarShip
