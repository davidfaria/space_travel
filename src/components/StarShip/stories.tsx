import { Story, Meta } from '@storybook/react/types-6-0'
import StarShip, { StartShipType } from '.'
import mock from './mock'

export default {
  title: 'StarShip',
  component: StarShip,
  args: mock
} as Meta

export const Default: Story<StartShipType> = () => (
  <StarShip starship={mock} distance={1000000} />
)
