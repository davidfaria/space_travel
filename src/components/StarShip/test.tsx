import { render, screen } from '@testing-library/react'
import mock from './mock'

import StarShip from '.'

describe('<StarShip />', () => {
  it('should render the starship', () => {
    const { container } = render(
      <StarShip distance={1000000} starship={mock} />
    )
    const imgShip = screen.getByRole('img', {
      name: /rocket ship/i
    })

    const imgStop = screen.getByRole('img', {
      name: /Stop Soard/i
    })

    expect(screen.getByText(/Millennium Falcon/)).toBeInTheDocument()
    expect(screen.getByText(/9/)).toBeInTheDocument()
    expect(imgShip).toBeInTheDocument()
    expect(imgStop).toBeInTheDocument()

    expect(imgShip).toHaveStyleRule('width', '32px')
    expect(imgStop).toHaveStyleRule('width', '48px')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the startship widthout distance prop', () => {
    const { container } = render(<StarShip starship={mock} />)
    expect(screen.getByText('*****')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the startship width consumables UNKNOW', () => {
    const mockUnknow = { ...mock, MGLT: undefined, consumables: 'unknow' }
    const { container } = render(
      <StarShip starship={mockUnknow} distance={1000000} />
    )

    expect(screen.getByText('*****')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
