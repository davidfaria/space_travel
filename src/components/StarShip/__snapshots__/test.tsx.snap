// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarShip /> should render the starship 1`] = `
.c0 {
  background: rgba(41,41,46,0.9);
  border-color: rgb(41,41,46);
  max-width: 380px;
  width: 100%;
  height: 100px;
  padding: 16px;
  margin-bottom: 8px;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.c0:hover {
  background: rgb(38,38,43);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 32px;
  height: auto;
  margin-right: 8px;
}

.c3 {
  color: #fff;
  font-weight: bold;
  font-size: 1.6rem;
  text-transform: uppercase;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  color: #fff;
  font-size: 2rem;
}

.c5 {
  width: 48px;
  height: auto;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Rocket Ship"
      class="c2"
      loading="lazy"
      src="/img/rocket.png"
    />
    <strong
      class="c3"
    >
      Millennium Falcon
    </strong>
  </div>
  <div
    class="c4"
  >
    <img
      alt="Stop Soard"
      class="c5"
      loading="lazy"
      src="/img/stop.png"
    />
    <span
      class="c6"
    >
      9
    </span>
  </div>
</div>
`

exports[
  `<StarShip /> should render the startship width consumables UNKNOW 1`
] = `
.c0 {
  background: rgba(41,41,46,0.9);
  border-color: rgb(41,41,46);
  max-width: 380px;
  width: 100%;
  height: 100px;
  padding: 16px;
  margin-bottom: 8px;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.c0:hover {
  background: rgb(38,38,43);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 32px;
  height: auto;
  margin-right: 8px;
}

.c3 {
  color: #fff;
  font-weight: bold;
  font-size: 1.6rem;
  text-transform: uppercase;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  color: #fff;
  font-size: 2rem;
}

.c5 {
  width: 48px;
  height: auto;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Rocket Ship"
      class="c2"
      loading="lazy"
      src="/img/rocket.png"
    />
    <strong
      class="c3"
    >
      Millennium Falcon
    </strong>
  </div>
  <div
    class="c4"
  >
    <img
      alt="Stop Soard"
      class="c5"
      loading="lazy"
      src="/img/stop.png"
    />
    <span
      class="c6"
    >
      *****
    </span>
  </div>
</div>
`

exports[`<StarShip /> should render the startship widthout distance prop 1`] = `
.c0 {
  background: rgba(41,41,46,0.9);
  border-color: rgb(41,41,46);
  max-width: 380px;
  width: 100%;
  height: 100px;
  padding: 16px;
  margin-bottom: 8px;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.c0:hover {
  background: rgb(38,38,43);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 32px;
  height: auto;
  margin-right: 8px;
}

.c3 {
  color: #fff;
  font-weight: bold;
  font-size: 1.6rem;
  text-transform: uppercase;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  color: #fff;
  font-size: 2rem;
}

.c5 {
  width: 48px;
  height: auto;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Rocket Ship"
      class="c2"
      loading="lazy"
      src="/img/rocket.png"
    />
    <strong
      class="c3"
    >
      Millennium Falcon
    </strong>
  </div>
  <div
    class="c4"
  >
    <img
      alt="Stop Soard"
      class="c5"
      loading="lazy"
      src="/img/stop.png"
    />
    <span
      class="c6"
    >
      *****
    </span>
  </div>
</div>
`
